src/client/datagram.cc

```diff
@@ -52,6 +52,41 @@ DatagramDestination::DatagramDestination(
     : m_Owner(owner),
       m_Receiver(nullptr) {}
 
+namespace {  // Helper facilities to transfer messages.
+
+/// Deleter for a shared pointer
+/// that acts like double handle.
+template <typename T>
+struct DeepDeleter {
+  /// Deletes the payload and handle.
+  void operator()(T** ptr) {
+    delete *ptr;
+    delete ptr;
+  }
+};
+
+/// Shared pointer with ``release`` semantics as unique_ptr.
+template <typename T>
+using ReleasableSharedPtr = std::shared_ptr<T*>;
+
+/// Helper function to make releasable shared_ptr with appropriate deleter.
+//
+/// @param ptr  The payload for the handle.
+template <typename T>
+auto make_releasable_shared_ptr(T* payload) {
+  return ReleasableSharedPtr<T>(new T*(payload), DeepDeleter<T>());
+}
+
+/// @returns The released payload from the shared ptr.
+template <typename T>
+T* release(ReleasableSharedPtr<T>& smart_ptr) {
+  T* paylaod = *smart_ptr;
+  *smart_ptr = nullptr;
+  return paylaod;
+}
+
+}  // namespace
+
```

Comment:
I tried to keep using ``unique_ptr`` but boost::asio started complaining that it required copy_constructable function. Moving functions didn't work.

I can refactor to make the msg work with shared pointers by default instead of unique_ptr.
it would remove some code.
However, this is a design choice.
Usually shared_ptr in places where unique_ptr will do is not a good sign.